driver-reactive-streams/src/test/functional/com/mongodb/reactivestreams/client/syncadapter/SyncMongoCursor.java

```diff
@@ -16,37 +16,138 @@
 
+import com.mongodb.MongoInterruptedException;
+import com.mongodb.MongoTimeoutException;
 import com.mongodb.ServerAddress;
 import com.mongodb.ServerCursor;
 import com.mongodb.client.MongoCursor;
 import com.mongodb.lang.Nullable;
 import org.reactivestreams.Publisher;
-import reactor.core.publisher.Flux;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
 
-import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.concurrent.BlockingDeque;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.TimeUnit;
+
+import static com.mongodb.ClusterFixture.TIMEOUT;
 
 class SyncMongoCursor<T> implements MongoCursor<T> {
-    private final Iterator<T> iterator;
+    private static final Object COMPLETED = new Object();
+    private final BlockingDeque<Object> results = new LinkedBlockingDeque<>();
+    private final Integer batchSize;
+    private int countToBatchSize;
+    private volatile Subscription subscription;
+    private T current;
+    private boolean completed;
+    private RuntimeException error;
+
+    SyncMongoCursor(final Publisher<T> publisher, @Nullable final Integer batchSize) {
+        this.batchSize = batchSize;
+        CountDownLatch latch = new CountDownLatch(1);
+        //noinspection ReactiveStreamsSubscriberImplementation
+        publisher.subscribe(new Subscriber<T>() {
+            @Override
+            public void onSubscribe(final Subscription s) {
+                subscription = s;
+                if (batchSize == null || batchSize == 0) {
+                    subscription.request(Long.MAX_VALUE);
+                } else {
+                    subscription.request(batchSize);
+                }
+                latch.countDown();
+            }
+
+            @Override
+            public void onNext(final T t) {
+                results.addLast(t);
+            }
 
-    SyncMongoCursor(final Publisher<T> publisher, final @Nullable Integer batchSize) {
-        iterator = (batchSize == null
-                ? Flux.from(publisher).toIterable()
-                : Flux.from(publisher).toIterable(batchSize))
-                .iterator();
+            @Override
+            public void onError(final Throwable t) {
+                results.addLast(t);
+            }
+
+            @Override
+            public void onComplete() {
+                results.addLast(COMPLETED);
+            }
+        });
+        try {
+            if (!latch.await(TIMEOUT, TimeUnit.SECONDS)) {
+                throw new MongoTimeoutException("Timeout waiting for subscription");
+            }
+            // Unfortunately this is the only to wait for the query to be initiated, since its asynchronous
+            // and we have no way of knowing
+            try {
+                Thread.sleep(100);
+            } catch (InterruptedException e) {
+                // ignore
```

Comment:
I decided to just throw MongoInterruptedException in all cases.

Reply:
Since the outer `try` has `catch (InterruptedException e)` we can just call `Thread.sleep` here without having a nested `try`, unless we actually care about differentiating whether `await` or `sleep` threw it (as far as I can tell, we do not care).

Reply:
I don't think we care either. 

Done